server/sonar-server/src/main/java/org/sonar/server/computation/queue/CeQueueImpl.java

```diff
@@ -253,4 +287,69 @@ public void resumePeek() {
   public boolean isPeekPaused() {
     return peekPaused.get();
   }
+
+  private static class CeQueueDtoToCeTask implements Function<CeQueueDto, CeTask> {
+    private final Map<String, ComponentDto> componentDtoByUuid;
+
+    public CeQueueDtoToCeTask() {
+      this.componentDtoByUuid = Collections.emptyMap();
+    }
+
+    public CeQueueDtoToCeTask(Map<String, ComponentDto> componentDtoByUuid) {
+      this.componentDtoByUuid = componentDtoByUuid;
+    }
+
+    @Override
+    @Nonnull
+    public CeTask apply(@Nonnull CeQueueDto dto) {
+      CeTask.Builder builder = new CeTask.Builder();
+      builder.setUuid(dto.getUuid());
+      builder.setType(dto.getTaskType());
+      builder.setSubmitterLogin(dto.getSubmitterLogin());
+      String componentUuid = dto.getComponentUuid();
+      if (componentUuid != null) {
+        builder.setComponentUuid(componentUuid);
+        ComponentDto component = componentDtoByUuid.get(componentUuid);
+        if (component != null) {
+          builder.setComponentKey(component.getKey());
+          builder.setComponentName(component.name());
+        }
+      }
+      return builder.build();
+    }
+  }
+
+  private static class CeTaskSubmitToInsertedCeQueueDto implements Function<CeTaskSubmit, CeQueueDto> {
+    private final DbSession dbSession;
+    private final DbClient dbClient;
+
+    public CeTaskSubmitToInsertedCeQueueDto(DbSession dbSession, DbClient dbClient) {
+      this.dbSession = dbSession;
+      this.dbClient = dbClient;
+    }
+
+    @Override
+    @Nonnull
+    public CeQueueDto apply(@Nonnull CeTaskSubmit submission) {
+      CeQueueDto dto = new CeQueueDto();
+      dto.setUuid(submission.getUuid());
+      dto.setTaskType(submission.getType());
+      dto.setComponentUuid(submission.getComponentUuid());
+      dto.setStatus(CeQueueDto.Status.PENDING);
+      dto.setSubmitterLogin(submission.getSubmitterLogin());
+      dto.setStartedAt(null);
+      dbClient.ceQueueDao().insert(dbSession, dto);
+      return dto;
+    }
+  }
+
+  private enum CeQueueDtoToComponentUuid implements Function<CeQueueDto, String> {
+    INSTANCE;
+
+    @Override
+    @Nullable
+    public String apply(@Nonnull CeQueueDto input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The "input" parameter nullability is different in the superclass method, and that should not be changed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)